src/compiler/protobuf_plugin.h

```diff
@@ -0,0 +1,261 @@
+   	 return GetModuleAndMessagePath(method_->output_type(), service,
+   									 out);
+   }
+
+   bool NoStreaming() const {
+     return !method_->client_streaming() && !method_->server_streaming();
+   }
+
+   bool ClientOnlyStreaming() const {
+     return method_->client_streaming() && !method_->server_streaming();
+   }
+
+   bool ClientStreaming() const {
+   	 return method_->client_streaming();
+   }
+
+   bool ServerOnlyStreaming() const {
+     return !method_->client_streaming() && method_->server_streaming();
+   }
+
+   bool ServerStreaming() const {
+   	 return method_->server_streaming();
+   }
+
+   bool BidiStreaming() const {
+     return method_->client_streaming() && method_->server_streaming();
+   }
+
+   grpc::string GetComments(bool leading, 
+                          const grpc::string prefix) const {
+     return GetCommentsHelper(method_, leading, prefix);
+   }
+
+   std::vector<grpc::string> GetAllComments() const {
+   	 return GetAllCommentsHelper(method_);
+   }
+
+  private:
+   const grpc::protobuf::MethodDescriptor *method_;
+   const grpc::protobuf::ServiceDescriptor *service;
+ };
+
+class ProtoBufService : public grpc_generator::Service {
+  public:
+   ProtoBufService(const grpc::protobuf::ServiceDescriptor *service)
+       : service_(service) {}
+
+   grpc::string name() const { return service_->name(); }
+
+   int method_count() const { return service_->method_count(); };
+   std::unique_ptr<const grpc_generator::Method> method(int i) const {
+     return std::unique_ptr<const grpc_generator::Method>(
+         new ProtoBufMethod(service_->method(i)));
+   };
+
+   grpc::string GetComments(bool leading, 
+                             const grpc::string prefix) const {
+     return GetCommentsHelper(service_, leading, prefix);
+   }
+
+   std::vector<grpc::string> GetAllComments() const {
+     return GetAllCommentsHelper(service_);
+   }
+
+  private:
+   const grpc::protobuf::ServiceDescriptor *service_;
+};
+
+class ProtoBufPrinter : public grpc_generator::Printer {
+  public:
+    ProtoBufPrinter(grpc::string *str)
+       : output_stream_(str), printer_(&output_stream_, '$') {}
+
+    void Print(const std::map<grpc::string, grpc::string> &vars,
+              const char *string_template) {
+      printer_.Print(vars, string_template);
+    }
+
+    void Print(const char *string) { printer_.Print(string); }
+    void Indent() { printer_.Indent(); }
+    void Outdent() { printer_.Outdent(); }
+
+  private:
+    grpc::protobuf::io::StringOutputStream output_stream_;
+    grpc::protobuf::io::Printer printer_;
+ };
+
+class ProtoBufFile : public grpc_generator::File {
+  public:
+   ProtoBufFile(const grpc::protobuf::FileDescriptor *file) : file_(file) {}
+
+   grpc::string filename() const { return file_->name(); }
+   grpc::string filename_without_ext() const {
+     return grpc_generator::StripProto(filename());
+   }
+
+   grpc::string message_header_ext() const { return ".pb.h"; }
```

Comment:
I think these two functions should be moved to language specific generators because the import/include statements will depend on the language.